gnovm/pkg/gnolang/values_string.go

```diff
@@ -103,49 +111,66 @@ func (av *ArrayValue) ProtectedString(seen *seenValues) string {
 	return fmt.Sprintf("array[0x%X]", av.Data)
 }
 
-func (sv *SliceValue) String() string {
-	return sv.ProtectedString(newSeenValues())
+func (sv *SliceValue) String(m *Machine) string {
+	return sv.ProtectedString(m, newSeenValues())
 }
 
-func (sv *SliceValue) ProtectedString(seen *seenValues) string {
+func (sv *SliceValue) ProtectedString(m *Machine, seen *seenValues) string {
+	m.GasMeter.ConsumeGas(OpCPUAssign, CPUCYCLES)
+
 	if sv.Base == nil {
 		return "nil-slice"
 	}
 
 	if seen.Contains(sv) {
+		m.GasMeter.ConsumeGas(OpCPUAssign, CPUCYCLES)
 		return fmt.Sprintf("%p", sv)
 	}
 
 	if ref, ok := sv.Base.(RefValue); ok {
+		m.GasMeter.ConsumeGas(OpCPUAssign, CPUCYCLES)
 		return fmt.Sprintf("slice[%v]", ref)
 	}
 
+	m.GasMeter.ConsumeGas(OpCPUAssign, CPUCYCLES)
 	seen.Put(sv)
 	defer seen.Pop()
 
 	vbase := sv.Base.(*ArrayValue)
 	if vbase.Data == nil {
+		m.GasMeter.ConsumeGas(int64(sv.Length*OpCPUAssign), CPUCYCLES)
 		ss := make([]string, sv.Length)
 		for i, e := range vbase.List[sv.Offset : sv.Offset+sv.Length] {
-			ss[i] = e.ProtectedString(seen)
+			m.GasMeter.ConsumeGas(OpCPUAssign, CPUCYCLES)
+			ss[i] = e.ProtectedString(m, seen)
 		}
+		m.GasMeter.ConsumeGas(int64(sv.Length*OpCPUAssign), CPUCYCLES)
 		return "slice[" + strings.Join(ss, ",") + "]"
 	}
+
+	defer func() {
+		m.GasMeter.ConsumeGas(OpCPUSlice, CPUCYCLES)
+	}()
+
```

Comment:
I'm not sure if this is the best solution at the moment. Perhaps we should develop a more precise model that can more accurately measure resource costs. For example, the gas cost for a slice value should roughly double if its size doubles. Does that seem more reasonable?

Reply:
Yea. I agree.